### ISSUE TITLE
Reducing usage of or even removing `groebner_assure`, `oscar_assure`, `singular_assure`

### ISSUE THREAD

**Issue by fingolfin** [oscar-system/Oscar.jl#1920]
In a recent PR, I saw code like this:
```julia
  groebner_assure(r.I, ordering)
  oscar_assure(r.I.gb[ordering])
  singular_assure(r.I.gb[ordering])
  SG = r.I.gb[ordering].gens.S
```
Why do we do it this way, instead of providing accessors for things like `I.gb[ordering].gens.S` that automatically do this for us? So let's say we
have a method 
```julia
function singular_groeber_gens(I::MPolyIdeal, ordering::MonomialOrdering)
  groebner_assure(r.I, ordering)
  singular_assure(r.I.gb[ordering])
  return I.gb[ordering].gens.S
end
```
and then original code above would just become
```julia
  SG = singular_groeber_gens(r.I, ordering)
```

Is there a fundamental problem with this alternative approach that I am missing?

Advantages of this approach include:
- we reduce the number of places that deal with implementation details (ideally most code would not need to know how and where Groebner bases are stored, and how they are computed
  -  this has many advantages: it makes future refactoring easier, makes it easier to write generic code, helps ensure code invariants ("if that value is set then this value also is set"), etc.
- the intent of the code is clearer: instead of randomly sprinkling `*_assure` calls, you just say what you want (e.g. in the example above, it is not readily clear whether and why `oscar_assure` is actually needed)
- the code arguably gets easier to read

**Comment by thofma** [comment]
Isn't this exactly what we did? The code you quoted is itself part of the `*_assure` function and it is called in the accessors here:
https://github.com/oscar-system/Oscar.jl/blob/169ccdf41f7ac1a55e828530253aa215b1a11534/src/Rings/MPolyQuo.jl#L55-L58

And everywhere these accessors are called, which is why this fundamental change in the behaviour only took a few lines in #1917.

Maybe one could save a few lines in total by refactoring it yet more, but why?

**Comment by ederc** [comment]
Yes, in the quotient business we are doing this. But @fingolfin is right, we should do this also in other settings, the direct call of `...gens.S` is widespread.

**Comment by fingolfin** [comment]
Yes, the example was from your PR and I realize that PR implemented a step in the right direction -- but I disagree that there are just a few lines to save; I did not target that PR specifically (and hence explicitly did not reference it), it just was a convenient source for an example. I could have picked another one of the hundreds which are remaining. Indeed:
```
$ git grep _assure src | fgrep -v 'function '  | wc -l
     181
```
Of course a few of those are accessors and fine, but the majority are not. Here are some examples I found by randomly picking grep matches:

```julia
function Base.:+(I::MPolyIdeal{T}, J::MPolyIdeal{T}) where T
  oscar_assure(I)
  oscar_assure(J)
  check_base_rings(I, J)
  newgens = filter!(!iszero, unique!(vcat(I.gens.O,J.gens.O)))
  return ideal(base_ring(I), newgens)
end
```
or
```julia
function saturation(I::MPolyIdeal{T}, J::MPolyIdeal{T}) where T
  singular_assure(I)
  singular_assure(J)
  K, _ = Singular.saturation(I.gens.S, J.gens.S)
  return MPolyIdeal(base_ring(I), K)
end
```
or
```julia
function Base.issubset(a::MPolyQuoIdeal{T}, b::MPolyQuoIdeal{T}) where T
  base_ring(a) == base_ring(b) || error("base rings must match")
  as = simplify(a)
  groebner_assure(b)
  return Singular.iszero(Singular.reduce(as.gens.S, b.gb.gens.S))
end
```
In each case, an accessor could (and IMHO should) be used which automatically takes care of computing the relevant data if it is missing. Perhaps those accessors already exist, though from a cursory look I couldn't find many (besides the ones you thankfully recently added).

Of course there might be cases where for various reasons we want to compute things explicitly up front (say to control how things are computed which may be more efficient if one sequences the steps in a certain way). But from a rough look, all examples I randomly sampled didn't seem to require that (of course I only looked at a small subset, and so perhaps I am wrong and there are tons of places where going via an accessor is a bad idea. But which are they?)

**Comment by thofma** [comment]
My "save a few lines" comment was with respect to the recent PR. I agree with what you say. This should be refactored.

**Comment by fingolfin** [comment]
@thofma sorry, seems I misunderstood :-/. Glad we agree then :-)

**Comment by fingolfin** [comment]
@jankoboehm will have a look at this in `src/Modules/UngradedModules/`

**Comment by fingolfin** [comment]
@JohnAAbbott maybe have a look at your `*_assure` functions in `src/Rings/PBWAlgebra.jl`

**Comment by fingolfin** [comment]
@jankoboehm will have another look at this.

And when @JohnAAbbott is back I hope he'll do it for PBW...

**Comment by JohnAAbbott** [comment]
> [@jankoboehm](https://github.com/jankoboehm) will have another look at this.
> 
> And when [@JohnAAbbott](https://github.com/JohnAAbbott) is back I hope he'll do it for PBW...

I have just looked though the code for `PBWAlgebraQuo`: there is only 1 call to an `assure` function (on line 131, in the implementation of `simplify`).  I'm unsure what might be a _clean and efficient_ way of eliminating it.  An idea might be to call `reduce` as defined on line 107 in `src/groebner/reduce.jl`  converting the generators of the `modulus` of the `PBWAlgQuo` ring by calling the constructor for `IdealGens` (line 168 or line 186 of `mpoly.jl`).  But this would seem to repeatedly create the `IdealGens` object (once for each call to `simplify`); also the result probably needs to be converted to Singular format.  Is this really better than the existing code?

**Comment by JohnAAbbott** [comment]
I'm now looking at `PBWAlgebra.jl`: there are several calls to `assure` functions here.  I'll see if any can reasonably be eliminated -- there is an obvious candidate near line 1043.  For the others, I must investigate...

**Comment by JohnAAbbott** [comment]
In `PBWAlgebra.jl` the functions `groebner_assure!` and `opgroebner_assure!` are defined on lines 717 and 731 respectively.   I think these implementations may contain a _race condition bug_: namely, if the user interrupts Julia **after** the field `gb` or `opgb` is set but **before** the subfield `isTwoSided` is set then the data-structure may be in an anomalous state.
One way to correct this would be to introduce yet another field, which is set only after the first two have been set; or maybe Julia has a feature for making some code blocks uninterruptible?
**UPDATE** I have just verified that a multiple assignment is not atomic in Julia (1.10): it can be interrupted when some values have been assigned but not others:
```
julia> a,b= 99,99;  while true   a,b = 0,0;  a,b = 1,2;  end
^C
julia> (a,b)
(1, 0)
```

**Comment by JohnAAbbott** [comment]
**In `PBWAlgebra.jl:`** We could remove some apparent/syntactic duplication in the functions `ideal_membership` and `is_subset`.  Otherwise I do not really see much opportunity to reduce the number of calls to `assure` functions.  I can try asking Hans for advice.

**Comment by lgoettgens** [comment]
> Otherwise I do not really see much opportunity to reduce the number of calls to `assure` functions.

I think the main purpose of this issue is not to completely remove all calls to `assure` functions, but instead move them into the innermost scope possible (and then also inline them there). Let me explain this at the example of 
https://github.com/oscar-system/Oscar.jl/blob/d92b966e0c29d2e5039fe5f7e04bb1c77ab3f245/src/Rings/PBWAlgebra.jl#L717-L725
with an example call to it in 
https://github.com/oscar-system/Oscar.jl/blob/d92b966e0c29d2e5039fe5f7e04bb1c77ab3f245/src/Rings/PBWAlgebra.jl#L821-L824

Instead of explicitly calling `groebner_assure!` on the ideal, and afterwards accessing its internal field `.gb`, it would instead be better to have some getter function for the `.gb` field that additionally also takes care of filling it with suitable data if needed.

The way I would approach it:
Add a function
```julia
function singular_groebner_basis(a::PBWAlgIdeal{D}) where D
  @assert D <= 0
  if !isdefined(a, :gb)
    a.gb = Singular.std(a.sdata)
    if D == 0
      a.gb.isTwoSided = false
    end
  end
  return a.gb
end
```
remove the function definition of `groebner_assure!` I mentioned above in this comment, and replacing the call to it (also see above in this comment) by
```julia
else 
  return _one_check(singular_groebner_basis(a)) 
end
```
and analogously for all other calls to `groebner_assure!`.

A similar pattern should also be possible to apply to most of the other assure-like functions.

**Comment by fingolfin** [comment]
@jankoboehm will take a look at the relevant code in `src/Modules`

**Comment by HechtiDerLachs** [comment]
There has been progress made within the last week. We have a look at what's left to do.

 * [ ] The occurrences in `examples/PrimDec.jl` should be removed.
 * [x] The same holds for `experimental/GITFans/src/GITFans.jl` (maybe already on the way somewhere?)
 * [ ] For the `MonoidAlgebra`s we need to get rid of the code duplication in total and in particular the `_assure` things.
 * [ ] For `ModStdX` @fieker and @ederc are still working on that (to move it out of `experimental`)
 * [ ] In `src/Modules/missing_functionality.jl` it can be removed (@HechtiDerLachs).
 * [ ] For `src/Modules/mpolyquo.jl` this is part of replacing that functionality with going directly to singular modules over quotient rings; see also #4816 
 * [x] Things in `src/Rings/NumberField.jl` are more in the yard of @fieker and @thofma ?
 * [ ] Stuff in `src/Rings/mpoly.jl` seem sane for the moment.
 * [x] @ederc will look into cleaning up `src/Rings/groebner/hilbert-driven.jl`.
 * [ ] there is a `groebner_assure` call in `test/Rings/groebner.jl:120`

**Comment by fingolfin** [comment]
There is also my PR #4937 which as a side effect also deals with some `*_assure` calls (replacing them). I'll rebase it on #4961 but it'd be nice if e.g. @ederc or @jankoboehm could review it.

**Comment by fingolfin** [comment]
I've ticked some of the things we already addressed in the list @ederc made last week -- I may have missed some.

I notice that `groebner_assure` has been essentially confined to a single file, `src/Rings/groebner/generators.jl`, and there really is only called in one place (the rest are "recursive" calls) -- with one exception, a call in `test/Rings/groebner.jl:120` which I hope can also be eliminated by someone.

**Comment by ederc** [comment]
Well, the question is if we want to get completely rid off the functionality of asking "Give me any GB for this ideal". Computing a GB via the other functions always means to specify a monomial ordering or to compute w.r.t. the default ordering.

**Comment by wdecker** [comment]
Asking "Give me any GB for this ideal" should be possible as this may save compting time

**Comment by fingolfin** [comment]
I don't understand why this is a question about whether "we want to get completely rid off the functionality of asking "Give me any GB for this ideal". 

To me having such a functionality is perfectly fine -- as long as it is a function that actually does that, i.e. `give_me_any_GB(I::Ideal)`.  

But `groebner_assure` does not do that -- it sets some internal field (`I.gens.gb` I believe?) to a GB.

**Comment by fingolfin** [comment]
Actually I consider this resolved now that `oscar_assure` and `singular_assure` are gone: yes `groebner_assure` lives but it is only used internally in a single file (and a call in a test file which I don't like but can live with). And it does return the object it computes. So it's OK by me to keep it like that.
